stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -243,10 +247,112 @@ swift::_matchMetadataByMangledTypeName(const llvm::StringRef typeName,
   return metadata;
 }
 
-// returns the type metadata for the type named by typeName
+// returns the generic type metadata for type named by boundGenericNode
+static const Metadata *
+matchMetadataGenericNode(const llvm::StringRef typeNodeName,
+                         const NodePointer boundGenericNode,
+                         const Metadata *unused __attribute__((unused)),
+                         const NominalTypeDescriptor *ntd) {
+  if (boundGenericNode->getNumChildren() != 2)
+    return nullptr;
+
+  auto templateTypeNode = boundGenericNode->getChild(0);
+  auto typeListNode = boundGenericNode->getChild(1);
+
+  if (templateTypeNode->getKind() != Node::Kind::Type ||
+      typeListNode->getKind() != Node::Kind::TypeList)
+    return nullptr;
+
+  // Validate that we have one or more generic parameters (the zero parameter
+  // case is handled by matchMetadataTypeNode) and that the number of primary
+  // archetypes matches the number of child nodes in the demangled type.
+  //
+  // TODO: support secondary archetypes, where NumParams != NumPrimaryParams;
+  // not sure if there is enough information available at runtime to do this.
+  auto &genericParams = ntd->GenericParams;
+  if (!genericParams.hasGenericParams() ||
+      genericParams.NumPrimaryParams != typeListNode->getNumChildren() ||
+      genericParams.NumParams != genericParams.NumPrimaryParams)
+    return nullptr;
+
+  auto templateName = mangleNode(templateTypeNode); // XXX cache this
+  if (ntd->Name.get() != templateName)
+    return nullptr;
+
+  auto pattern = ntd->getGenericMetadataPattern();
+  void **arguments = new void *[pattern->NumKeyArguments];
+
+  for (size_t i = 0, witnessIndex = 0;
+       i < genericParams.NumPrimaryParams; i++) {
+    auto param = genericParams.getParameterAt(i);
+    auto metadata = findMetadataForNode(typeListNode->getChild(i));
+    if (metadata == nullptr) {
+      // unknown type
+      delete arguments;
+      return nullptr;
+    }
+
+    for (size_t j = 0; j < param->NumWitnessTables; j++, witnessIndex++) {
+      auto &protocol = param->Protocols[j];
+      auto wt = swift_conformsToProtocol(metadata, protocol);
+      if (wt == nullptr) {
+        // does not conform to associated type requirement, bail
+        delete arguments;
+        return nullptr;
+      }
+
+      arguments[genericParams.NumPrimaryParams + witnessIndex] =
+        const_cast<void *>(reinterpret_cast<const void *>(wt));
+    }
+
+    arguments[i] = const_cast<void *>(reinterpret_cast<const void *>(metadata));
+  }
+
+  auto metadata = swift_getGenericMetadata(pattern, arguments);
```

Comment:
Is it possible for swift_getGenericMetadata() to ever take the section scan lock? If so you will have a potential deadlock here even if the section lock is recursive because of lock ordering between threads.